src/zip.c

```diff
@@ -950,10 +934,88 @@ int zip_extract(const char *zipname, const char *dir,
 
 out:
   // Close the archive, freeing any resources it was using
-  if (!mz_zip_reader_end(&zip_archive)) {
+  if (!mz_zip_reader_end(zip_archive)) {
     // Cannot end zip reader
     status = -1;
   }
+  return status;
+
+}
+
+static mz_zip_archive * zip_archive_init_file_reader(const char *zipname, mz_uint32 flags){
+  mz_zip_archive *zip_archive = NULL;
+  zip_archive = (mz_zip_archive *)malloc(sizeof(mz_zip_archive));
+  if (!zip_archive){
+    // malloc failed.
+    return NULL;
+  }
+  if (!memset(zip_archive, 0, sizeof(mz_zip_archive))) {
+    // Cannot memset zip archive
+    return NULL;
+  }
+  // Now try to open the archive.
+  if (!mz_zip_reader_init_file(zip_archive, zipname, flags)) {
+    // Cannot initialize zip_archive reader
+    return NULL;
+  }
+  return zip_archive;
+}
+
+int zip_extract(const char *zipname, const char *dir,
+                 int (*on_extract)(const char *filename, void *arg), void *arg) {
+  if (!zipname || !dir) {
+    // Cannot parse zip archive name
+    return -1;
+  }
+  // init zip_archive and set reader
+  mz_zip_archive *zip_archive = zip_archive_init_file_reader(zipname, 0);
+  if (!zip_archive){
+    return -1;
+  }
+
+  int status = zip_extract_internal(zip_archive, dir, on_extract, arg);
+
+  free(zip_archive);
 
   return status;
 }
+
+static mz_zip_archive * zip_archive_init_mem_reader(const void *stream,
+                                      size_t size, mz_uint32 flags){
+  mz_zip_archive *zip_archive = NULL;
+  zip_archive = (mz_zip_archive *)malloc(sizeof(mz_zip_archive));
+  if (!zip_archive){
+    // malloc failed.
+    return NULL;
+  }
+  if (!memset(zip_archive, 0, sizeof(mz_zip_archive))) {
+    // Cannot memset zip archive
+    return NULL;
```

Comment:
See comments above regarding `malloc` and `free` if we fail  after successful `malloc`

Reply:
ok.